---
title: "HW4_TCGA"
output: html_document
editor_options: Assaf Lovtov & Adi Falach
chunk_output_type: console
---

```{r}
library(survival)
library(survminer)
```

# Analyzing TCGA data using the RTCGA package

The RTCGA package (bioconductor.org/packages/RTCGA) and all the associated data packages provide convenient access to clinical and genomic data in TCGA. Each of the data packages is a separate package, and must be installed (once) individually.

```{r}

#BiocManager::install("RTCGA")
#BiocManager::install("RTCGA.clinical") # if asked, do not install from source.
#BiocManager::install("RTCGA.mRNA") # if asked, do not install from source.

library("RTCGA.mRNA")

```

Let’s load the RTCGA package, and use the infoTCGA() function to get some information about the kind of data available for each cancer type.

```{r}
library(RTCGA) 
infoTCGA()
```

Next, let’s load the RTCGA.clinical package and get a little help about what’s available there.

```{r echo=FALSE}
library("RTCGA.clinical")
?clinical
```

This tells us all the clinical datasets available for each cancer type. If we just focus on breast cancer, look at how big the data is! There are 1098 rows by 3703 columns in this data alone. Let’s look at some of the variable names. Be careful with View() here – with so many columns, depending on which version of RStudio you have that may or may not have fixed this issue, Viewing a large dataset like this may lock up your RStudio.

```{r echo=FALSE}
dim(BRCA.clinical)
names(BRCA.clinical)
```

We’re going to use the survivalTCGA() function from the RTCGA package to pull out survival information from the clinical data. It does this by looking at vital status (dead or alive) and creating a times variable that’s either the days to death or the days followed up before being censored. Look at the help for ?survivalTCGA for more info. You give it a list of clinical datasets to pull from, and a character vector of variables to extract. Let’s look at breast cancer, ovarian cancer, and glioblastoma multiforme. Let’s just extract the cancer type (admin.disease_code).

```{r}
# Create the clinical data
clin <- survivalTCGA(BRCA.clinical, OV.clinical, GBM.clinical, 
                     extract.cols="admin.disease_code")
# Show the first few lines
head(clin)
```

```{r}
# How many samples of each type?
table(clin$admin.disease_code)
```

```{r}
# Tabulate by outcome
xtabs(~admin.disease_code+patient.vital_status, data=clin) %>% addmargins()
```

Now let’s run a Cox PH model against the disease code. By default it’s going to treat breast cancer as the baseline, because alphabetically it’s first. But you can reorder this if you want with factor().

```{r}

coxph(Surv(times, patient.vital_status)~admin.disease_code, data=clin)

```

This tells us that compared to the baseline brca group, GBM patients have a ~18x increase in hazards, and ovarian cancer patients have ~5x worse survival. Let’s create a survival curve, visualize it with a Kaplan-Meier plot, and show a table for the first 5 years survival rates.


```{r}

sfit <- survfit(Surv(times, patient.vital_status)~admin.disease_code, data=clin)
summary(sfit, times=seq(0,365*5,365))
```

```{r}
ggsurvplot(sfit, conf.int=TRUE, pval=TRUE)
```

Let’s load the gene expression data.

```{r}
library(RTCGA.mRNA)
?mRNA
```

Take a look at the size of the BRCA.mRNA dataset, show a few rows and columns.

```{r}
dim(BRCA.mRNA)
BRCA.mRNA[1:5, 1:5]
```

Similar to how survivalTCGA() was a nice helper function to pull out survival information from multiple different clinical datasets, expressionsTCGA() can pull out specific gene expression measurements across different cancer types. See the help for ?expressionsTCGA. Let’s pull out data for PAX8, GATA-3, and the estrogen receptor genes from breast, ovarian, and endometrial cancer, and plot the expression of each with a box plot.

```{r}
library(ggplot2)
expr <- expressionsTCGA(BRCA.mRNA, OV.mRNA, UCEC.mRNA,
                        extract.cols = c("PAX8", "GATA3", "ESR1"))
head(expr)
```

```{r}
table(expr$dataset)
```

```{r}
ggplot(expr, aes(dataset, PAX8, fill=dataset)) + geom_boxplot()
```

```{r}
ggplot(expr, aes(dataset, GATA3, fill=dataset)) + geom_boxplot()
```

```{r}
ggplot(expr, aes(dataset, ESR1, fill=dataset)) + geom_violin()
```

Or all together:

```{r}
library(tidyr)
expr %>% 
  as_tibble() %>% 
  gather(gene, expression, PAX8, GATA3, ESR1) %>% 
  ggplot(aes(dataset, expression, fill=dataset)) + 
    geom_boxplot() + 
    facet_wrap(~gene)
```

# Question 3

The “KIPAN” cohort (in KIPAN.clinical) is the pan-kidney cohort, consisting of KICH (chromaphobe renal cell carcinoma), KIRC (renal clear cell carcinoma), and KIPR (papillary cell carcinoma). The KIPAN.clinical has KICH.clinical, KIRC.clinical, and KIPR.clinical all combined.

1. Using survivalTCGA(), create a new object called `clinkid` using the KIPAN.clinical cohort. For the columns to extract, get both the disease code and the patient's gender (extract.cols=c("admin.disease_code", "patient.gender")). 

```{r}
clinkid <- survivalTCGA(KIPAN.clinical, extract.cols=c("admin.disease_code", "patient.gender"))

head(clinkid)

```
2. The xtabs() command will produce tables of counts for categorical variables. Here’s an example for how to use xtabs() for the built-in colon cancer dataset, which will tell you the number of samples split by sex and by treatment.
```{r}
xtabs(~rx+sex, data=colon)
```

Use the same command to examine how many samples you have for each kidney sample type, separately by sex.
```{r, eval=FALSE}
xtabs(~admin.disease_code+patient.gender, data=clinkid)
```

3. Run a Cox PH regression on the cancer type and gender. What’s the effect of gender? Is it significant? How does survival differ by each type? Which has the worst prognosis?

Answer:

As we can see we got a high p value for the Cox PH regression on the gender that means it is not significant. On the other hand, we got a p value smaller than 0.05
for the cox regression on the cancer type which means it is significant.

Compared to the baseline group, kric patients have a ~5x increase in hazards, and kirp cancer patients have ~3x worse survival.Therefore we deduce that kirp patients have a 
better prognosis.


```{r, eval=FALSE}
coxph(Surv(times, patient.vital_status)~admin.disease_code+patient.gender, data=clinkid)
```


4. Create survival curves for each different subtype.
+ Produce a Kaplan-Meier plot.
+ Show survival tables each year for the first 5 years.
```{r, eval=FALSE}
sfit <- survfit(Surv(times, patient.vital_status)~admin.disease_code, data=clinkid)
summary(sfit, times=seq(0,365*5,365))
ggsurvplot(sfit, conf.int=TRUE, pval=TRUE)
```

5. Now we will add the gene expression data.

+ Extract gene expression data of all genes for the KIPAN dataset and merge this data with the
clinical data.

```{r}
exprkid <- expressionsTCGA(KIPAN.mRNA)
```
The patient barcode (bcr_patient_barcode) is not similar in both the clinical and expression data.
Thus, to merge the dataset you will first need to use a substring from the bcr_patient_barcode at the expression dataset.

```{r, eval=FALSE}
exprkid$bcr_patient_barcode = substr(exprkid$bcr_patient_barcode,0,12)
```

and then you may use the merge function:

```{r, eval=FALSE}
m = merge(clinkid,exprkid,by='bcr_patient_barcode')
colnames(m) <- make.names(colnames(m))
```

Find the gene that its expression is mostly associated (lowest p-value) with prognosis. 

Hint 1: Use a loop to go over all genes, and in each iteration use the coxph function with the expression levels of each gene.

Hint 2: To extract the p-value use the `summary` function on the output of the coxph analysis.
The p-values are within the `coefficient` filed in of `summary`.
```{r}
min_gene <- ''
min_p <- 1

for (gene in names(m)[6:17819]) 
  {
  sum <- summary(coxph(Surv(times, patient.vital_status)~get(gene), data=m))
  p <- sum$coefficients[,5]
  if (min_p > p)
  {
    min_gene <- gene
    min_p <- p
  }
}
min_gene

```



Plot the KM-plot by stratifying the patients by the median expression of the gene you found.

```{r}
library(dplyr)
print(min_gene)
gene_med <- median(m$HERC3)

lower <- filter(m, HERC3 < gene_med)
higher <- filter(m, HERC3 >= gene_med)

lower$patient_above_median <- 'false'
higher$patient_above_median <- 'true'

new_m <- rbind(lower, higher)
ggsurvplot(surv_fit(Surv(times, patient.vital_status)~patient_above_median, data=new_m))
```


